File: code/queens.py
#!/usr/bin/env python3
#-----------------------------------------------------------------------------
#                         
#=============================================================================

#%---------------------------------------------------------------------------
#                                IMPORTS
#-----------------------------------------------------------------------------
#%%

import numpy as np
import numba
from   numba import njit,prange

from timeit    import timeit
from functools import partial
from multiprocessing import Pool as Pool_proc

from util import addToFile
from math import sqrt
#%%---------------------------------------------------------------------------
#                            SEQUENTIAL          
#-----------------------------------------------------------------------------

def allQueensRec(n):   
    # Arrays used to flag available columns and diagonals
    col = np.ones(n,dtype=bool)
    dg1 = np.ones(2*n,dtype=bool)
    dg2 = np.ones(2*n,dtype=bool)
    
    return allQueensAux(n,0,col,dg1,dg2)
    
def allQueensAux(n,i,col,dg1,dg2):
    # All rows are filled. Stop the recursion.
    if n == i : 
        return 1
    # Try putting a queen in each cell of row i 
    nsol = 0
    for j in range(n):
        if (col[j] and dg1[i+j] and dg2[i-j+n]): 
            
            col[j]     = False  # Mark column j as occupied
            dg1[i+j]   = False  # Mark diagonal i+j as occupied
            dg2[i-j+n] = False  # Mark diagonal i-j as occupied

            nsol+=allQueensAux(n,i+1,col,dg1,dg2)
            
            col[j]     = True   # Unmark column j
            dg1[i+j]   = True   # Unmark diagonal i+j
            dg2[i-j+n] = True   # Unmark diagonal i-j
            
    return nsol

if __name__ == "__main__":
    ns = allQueensRec(8)
    print('Simple :',ns)
#%%
@njit(nogil=True) #Numba decorator
def allQueensNmb(n,i=0,col=None,dg1=None,dg2=None):
    # The array types must be numba compatible
    col = np.ones(n,dtype=numba.boolean)
    dg1 = np.ones(2*n,dtype=numba.boolean)
    dg2 = np.ones(2*n,dtype=numba.boolean)
    
    return allQueensNumba(n,0,col,dg1,dg2)
    
@njit(nogil=True) #Numba decorator
def allQueensNumba(n,i,col,dg1,dg2):
    # All rows are filled. Stop the recursion
    if n == i :
        return 1
    # Try putting a queen in each cell of row i
    nsol=0
    for j in range(n):
        if (col[j] and dg1[i+j] and dg2[i-j+n]): 

            col[j]     = False  # Mark column j as occupied
            dg1[i+j]   = False  # Mark diagonal i+j as occupied
            dg2[i-j+n] = False  # Mark diagonal i-j as occupied

            nsol+=allQueensNumba(n,i+1,col,dg1,dg2)
            
            col[j]     = True   # Unmark column j
            dg1[i+j]   = True   # Unmark diagonal i+j
            dg2[i-j+n] = True   # Unmark diagonal i-j
            
    return nsol

if __name__ == "__main__":
    ns = allQueensNmb(8)
    print('Numba  :',ns)
#%%---------------------------------------------------------------------------
#                            PARALLEL         
#-----------------------------------------------------------------------------
#%%
@njit(nogil=True)   
def allQueensCol(n,j):
    
    col = np.ones(n,dtype=numba.boolean)
    dg1 = np.ones(2*n,dtype=numba.boolean)
    dg2 = np.ones(2*n,dtype=numba.boolean)
    # Put a queen in cell j of the first row
    col[j] = False
    dg1[j] = False
    dg2[n-j] = False
    # Fills the rest of the board starting with the second row
    return allQueensNumba(n,1,col,dg1,dg2)

    
if __name__ == "__main__":
    nsol = 0
    for j in range(8):
        nsol += allQueensCol(8,j)
    print('Par    :',nsol)
#%%
@njit(nogil=True,parallel=True)
def allQueensPara(n):
    nsol = 0
    for j in prange(n):
        nsol+=allQueensCol(n,j)
    return nsol

#@njit()
def poolWorker(n,j):
    return allQueensCol(n,j)

def allQueensPool(n,np=None):
    if(np is None):
        with Pool_proc() as pool:
            nsols= pool.map(partial(poolWorker,n),prange(n))
            return (sum(nsols))
    else:
        with Pool_proc(processes=np) as pool:
            nsols= pool.map(partial(poolWorker,n),prange(n))
            return (sum(nsols))
    
if __name__ == "__main__":
    print('Seq    :' ,allQueensPara(8))
    print('Pool   :',allQueensPool(8,None))
#%%---------------------------------------------------------------------------
#                            TEST
#-----------------------------------------------------------------------------
#%%
def timeit_and_print(statement,number=10, name=''):
    time_elapsed = timeit(statement,number=number)
    
    time_per_iter = time_elapsed/number
    
    print('{ti:04f} s/iter | {tt:04f} s total | {name}'.format(
        ti = time_per_iter,
        tt = time_elapsed,
        name = name,
    ))
    return time_per_iter

def timeit_and_var(statement,number=10, name=''):
    
    if(number<1):
        return 0.0,0.0
    
    mean = 0.0
    var  = 0.0
    
    for i in range(number):
        dt = timeit(statement,number=1)
        mean += dt
        var  += dt*dt
        
    numb = float(number)
    mean = mean / numb
    var  = var  / numb  - mean*mean
    var *= (numb / (numb - 1.0))
    var  = sqrt(var)
        
    
    print('{ti:04f} s/iter | {tt:04f} variance | {name}'.format(
        ti = mean,
        tt = var,
        name = name,
    ))
    return mean,var
#%%
print('----------------------------------------------------------')
for n in range(8,19):
    rpt1=0
    rpt2=0
    rpt3=0
    rpt4=0
    if(n<15):
        rpt1=20
        rpt2=20
        rpt3=20
        rpt4=20
    elif(n<17):
        rpt2=10
        rpt3=10
        rpt4=10
    else:
        rpt3=3
        rpt4=3
        
    rpt1 = 0
    
    m1,v1=timeit_and_var(partial(allQueensRec,n),  name='Sequential python {:2d}'.format(n),number=rpt1) 
    m2,v2=timeit_and_var(partial(allQueensNmb,n),  name='Sequential numba  {:2d}'.format(n),number=rpt2)
    m3,v3=timeit_and_var(partial(allQueensPara,n), name='Parallel  para    {:2d}'.format(n),number=rpt3)
    m4,v4=timeit_and_var(partial(allQueensPool,n), name='Parallel  pool    {:2d}'.format(n),number=rpt4)
    addToFile('/Users/fua/tmp/python.var','{:2d} {:f} {:f} {:f} {:f} {:f} {:f} {:f} {:f}\n'.format(n,m1,v1,m2,v2,m3,v3,m4,v4),newP=(8==n))
    print('----------------------------------------------------------')
#%%
print('----------------------------------------------------------')
rpt = 0
if(rpt>0):
    for n in range(15,19):
        m1,v1=0.0
        m2.v2=0.0
        m3,v3=timeit_and_var(partial(allQueensPara,n), name='Parallel  para    {:2d}'.format(n),number=rpt)
        m4,v4=timeit_and_var(partial(allQueensPool,n), name='Parallel  pool    {:2d}'.format(n),number=rpt)
        addToFile('/Users/fua/tmp/python.var','{:2d} {:f} {:f} {:f} {:f} {:f} {:f} {:f} {:f}\n'.format(n,m1,v1,m2,v2,m3,v3,m4,v4),newP=False)
#%%
print('----------------------------------------------------------')
rpt = 0
if(rpt>0):
    for n in range(15,19):
        dt1=0.0
        dt2=0.0
        dt3=0.0
        dt4=timeit_and_print(partial(allQueensPool,n), name='Parallel  pool    {:2d}'.format(n),number=rpt)
        addToFile('/Users/fua/tmp/python.dat','{:2d} {:f} {:f} {:f} {:f}\n'.format(n,dt1,dt2,dt3,dt4),newP=False)
        print('----------------------------------------------------------')
#%%
print('----------------------------------------------------------')
rpt = 0
if(rpt>0):
    for n in range(8,14):
        dt1=0.0
        dt2=0.0
        dt3=timeit_and_print(partial(allQueensPara,n), name='Parallel  para    {:2d}'.format(n),number=rpt)
        dt4=0.0
        addToFile('/Users/fua/tmp/python.dat','{:2d} {:f} {:f} {:f} {:f}\n'.format(n,dt1,dt2,dt3,dt4),newP=False)
        print('----------------------------------------------------------')
#%%
#number  = 5
#mean = 0.0
#var  = 0.0
#ar = np.random.randn(number)
#for dt in ar:
#    mean += dt
#    var  += dt*dt
#    
#numb = float(number)
#mean = mean / numb
#var  = var  / numb  - mean*mean
#var *= (numb / (numb - 1.0))
#var  = sqrt(var)
#print(mean,var)

File: src/Queens.java

class Queens {
	
	public static void main(String[] args){
		
		// b.col = new boolean[n];
		// b.dg1 = new boolean[2*n];
		// b.dg2 = new boolean[2*n];
		
		// for (int i=0;i<2*n;i++){
		// 	b.dg1[i] = true;
		// 	b.dg2[i] = true;	
		// }
		// for (int i=0;i<n;i++){
		// 	b.col[i] = true;	
		// }
		
		//System.out.print("n= " + n + ", found " + nsol + " solutions.\n");

		System.out.println("n	mean	var");

		for (int num_queens = 8; num_queens <= 16; num_queens++) {
			int num_iter = 5;
			if(num_queens <= 10) {
				num_iter = 1000;
			} else if (num_queens <= 15) {
				num_iter = 10;
			}

			int nsol = 0;
			double mean = 0;
			double var = 0;


			for (int iter_idx = 0; iter_idx < num_iter; iter_idx++) {

				double t1 = System.nanoTime();

				Board b = new Board(num_queens);
				nsol = b.allQueensRec();

				double t2 = System.nanoTime();
				double dt = (t2 - t1) * 1e-9;

				mean += dt;
				var += dt*dt;
			}

			mean /= num_iter;

			if(num_iter > 1){
				double num_iter_d = num_iter;
				var = var  / num_iter_d  - mean*mean;
				var *= (num_iter_d / (num_iter_d - 1.0));
				var = Math.sqrt(var);
			} else {
				var = 0;
			}

			//System.out.println(String.format("%d %d", num_queens, nsol));

		
			// println(s"$num_queens queens, $nsol solutions time $t_mean s per attempt. (var $t_var, nt $num_iter)")
			System.out.println(String.format("%d	%f	%f", num_queens, mean, var));
		}



	}
	
}


class Board {

	protected int n;
	protected boolean col[],dg1[],dg2[];

	public Board(int n) {
		this.n = n;
		col = new boolean[n];
		dg1 = new boolean[2*n];
		dg2 = new boolean[2*n];
	}

	int addQueensRecursive(int n, int i) {

		if (n == i) {
			return 1;
		}

		int nsol = 0;
		for (int j = 0; j < n; j++) {
			if (col[j] && dg1[i+j] && dg2[i-j+n]) {
				col[j] = false;
				dg1[i+j] = false;
				dg2[i-j+n] = false;

				nsol += addQueensRecursive(n, i+1);

				col[j] = true;
				dg1[i+j] = true;
				dg2[i-j+n] = true;
			}
		}
		return nsol;
	}

	public int allQueensRec() {
		for (int i=0;i<2*n;i++){
			dg1[i] = true;
			dg2[i] = true;	
		}
		for (int i=0;i<n;i++){
			col[i] = true;	
		}
		return addQueensRecursive(n, 0);
	}
}

